sdk/storage/azure-storage-blob-nio/src/main/java/com/azure/storage/blob/nio/AzureDirectoryStream.java

```diff
@@ -0,0 +1,188 @@
+
+            BlobContainerClient containerClient;
+            ListBlobsOptions listOptions = new ListBlobsOptions()
+                .setDetails(new BlobListDetails().setRetrieveMetadata(true));
+            if (path.isRoot()) {
+                String containerName = path.toString().substring(0, path.toString().length() - 1);
+                containerClient = ((AzureFileSystem) path.getFileSystem()).getBlobServiceClient()
+                    .getBlobContainerClient(containerName);
+            } else {
+                AzureResource azureResource = new AzureResource(path);
+                listOptions.setPrefix(azureResource.getBlobClient().getBlobName() + AzureFileSystem.PATH_SEPARATOR);
+                containerClient = azureResource.getContainerClient();
+            }
+            this.blobIterator = containerClient
+                .listBlobsByHierarchy(AzureFileSystem.PATH_SEPARATOR, listOptions, null).iterator();
+        }
+
+        @Override
+        public boolean hasNext() {
+            // Closing the parent stream halts iteration.
+            if (parentStream.closed) {
+                return false;
+            }
+
+            // In case a customer calls hasNext multiple times in a row. If we've buffered an element, we have a next.
+            if (this.bufferedNext != null) {
+                return true;
+            }
+
+            /*
+            Search for a new element that passes the filter and buffer it when found. If no such element is found,
+            return false.
+             */
+            while (this.blobIterator.hasNext()) {
+                BlobItem nextBlob = this.blobIterator.next();
+                Path nextPath = getNextListResult(nextBlob);
+                try {
+                    if (this.filter.accept(nextPath) && passesDirectoryDuplicateFilter(nextPath, nextBlob)) {
+                        this.bufferedNext = nextPath;
+                        return true;
+                    }
+                } catch (IOException e) {
+                    throw LoggingUtility.logError(logger, new DirectoryIteratorException(e));
+                }
+            }
+            return false;
+        }
+
+        @Override
+        public Path next() {
+            if (this.bufferedNext == null) {
+                if (!this.hasNext()) { // This will populate cachedNext in the process.
+                    throw LoggingUtility.logError(logger, new NoSuchElementException());
+                }
+            }
+            Path next = this.bufferedNext; // cachedNext will have been populated by hasNext()
+            this.bufferedNext = null;
+            return next;
+        }
+
+        @Override
+        public void remove() {
+            throw new UnsupportedOperationException();
+        }
+
+        private Path getNextListResult(BlobItem blobItem) {
+            // Strip the root if it is present so we can relativize the list result, which never has a root.
+            Path withoutRoot = this.path;
```

Comment:
is there a reason we can't just do this +next 3 lines once in the constructor for AzureDirectoryIterator?

Reply:
Ooo probably. Save doing it every time. Good thought!